sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -248,7 +250,8 @@ case class StaticInvoke(
     arguments: Seq[Expression] = Nil,
     inputTypes: Seq[AbstractDataType] = Nil,
     propagateNull: Boolean = true,
-    returnNullable: Boolean = true) extends InvokeLike {
+    returnNullable: Boolean = true,
+    isDeterministic: Boolean = true) extends InvokeLike {
```

Comment:
I think this might be controversial to backport .. as
1. V2 expressions are unstable yet.
2. It could lead to different results in maintenance version upgrade if a user sets `isDeterministic` to `false`
3. Maybe performance regression if a user sets `isDeterministic` to `false`.
4. We haven't heard an actual complaint from user mailing list.
5. This makes an API (`isDeterministic`) working that has never been working in a way (is this a bug fix?)

While I agree with this being merged in 3.3.0, and I don't feel strongly on this, maybe we can consider reverting this out of `branch-3.2` because it has a good and bad thing. If we're worried about this change, we could issue a warning instead when `isDeterministic` from V2 Scalar Function returns `false`.

I will leave it to you @sunchao.

Reply:
I think this is a bug fix. A V2 catalog can return a function that's non-deterministic, while without the fix Spark can treat it as deterministic and apply related optimization rules (e.g., constant folding), which could cause correctness issues. 

Since this is already in Spark 3.2.1, I don't see much benefit of reverting it and re-introduce the correctness issue.

Reply:
+1 on keeping this in 3.2.x. This fixed the correctness issue and we actually intentionally included this fix in 3.2.1 release.

Reply:
That's fine. I didn't have a strong preference so I am okay with keeping it either 👍 .